addons/utm/models/utm_mixin.py

```diff
@@ -59,3 +55,29 @@ def tracking_fields(self):
             ('utm_source', 'source_id', 'odoo_utm_source'),
             ('utm_medium', 'medium_id', 'odoo_utm_medium'),
         ]
+
+    def _find_or_create_record(self, model_name, name):
+        """Based on the URL parameter, retrieve the corresponding record or create it."""
+        Model = self.env[model_name]
+
+        record = Model.search([('name', 'ilike', name)], limit=1)
+
+        if not record:
+            # No record found, create a new one
+            record_values = {'name': name}
+            if 'is_auto_campaign' in record._fields:
+                record_values['is_auto_campaign'] = True
+            record = Model.create(record_values)
+
+        return record
+
+    @api.model
+    def _generate_unique_name(self, record, name):
```

Comment:
Would say ``_set_name_unique`` as it does not generate a name, just ensure it is unique. I would also not give record but the model actually.